src/Management/src/Endpoint/Actuators/HttpExchanges/HttpExchangesDiagnosticObserver.cs

```diff
@@ -0,0 +1,202 @@
+        {
+            RecordHttpTrace(current, context);
+        }
+    }
+
+    private HttpExchange GetHttpExchange(HttpContext context, TimeSpan duration)
+    {
+        HttpExchangesEndpointOptions options = _optionsMonitor.CurrentValue;
+        string requestUri = GetRequestUri(context.Request, options.AddPathInfo);
+
+        if (options.AddQueryString)
+        {
+            requestUri += context.Request.QueryString.Value;
+        }
+
+        string? remoteAddress = null;
+
+        if (options.AddRemoteAddress)
+        {
+            remoteAddress = GetRemoteAddress(context);
+        }
+
+        Dictionary<string, IList<string?>> requestHeaders = [];
+
+        if (options.AddRequestHeaders)
+        {
+            requestHeaders = GetHeaders(context.Request.Headers, options.RequestHeaders);
+        }
+
+        var traceRequest = new TraceRequest(context.Request.Method, requestUri, requestHeaders, remoteAddress);
+
+        Dictionary<string, IList<string?>> responseHeaders = [];
+
+        if (options.AddResponseHeaders)
+        {
+            responseHeaders = GetHeaders(context.Response.Headers, options.ResponseHeaders);
+        }
+
+        var traceResponse = new TraceResponse(context.Response.StatusCode, responseHeaders);
+
+        string? userName = null;
+
+        if (options.AddUserPrincipal)
+        {
+            userName = GetUserPrincipal(context);
+        }
+
+        TracePrincipal? principal = userName == null ? null : new TracePrincipal(userName);
+
+        string? sessionId = null;
+
+        if (options.AddSessionId)
+        {
+            sessionId = GetSessionId(context);
+        }
+
+        TraceSession? session = sessionId == null ? null : new TraceSession(sessionId);
+
+        long timestamp = GetJavaTime(DateTime.UtcNow.Ticks);
+
+        return new HttpExchange(traceRequest, traceResponse, timestamp, principal, session, duration.TotalMilliseconds);
+    }
+
+    internal static HttpContext? GetHttpContextPropertyValue(object instance)
+    {
+        return GetPropertyOrDefault<HttpContext>(instance, "HttpContext");
+    }
+
+    internal static Dictionary<string, IList<string?>> GetHeaders(IHeaderDictionary headers, IList<string> allowedHeaders)
+    {
+        var result = new Dictionary<string, IList<string?>>();
+
+        foreach (KeyValuePair<string, StringValues> pair in headers)
+        {
+#pragma warning disable S4040 // Strings should be normalized to uppercase
+            result.Add(pair.Key.ToLowerInvariant(), HeaderShouldBeRedacted(pair.Key, allowedHeaders) ? new StringValues(Redacted) : pair.Value);
+#pragma warning restore S4040 // Strings should be normalized to uppercase
+        }
+
+        return result;
+    }
+
+    private static bool HeaderShouldBeRedacted(string currentHeader, IList<string> allowedHeaders)
+    {
+        return !allowedHeaders.Contains(currentHeader) || (allowedHeaders.Count == 1 && string.IsNullOrEmpty(allowedHeaders[0]));
```

Comment:
Is the right-hand side of the `||` expression reachable? Because the list gets cleared from `ConfigureOptions`.

Reply:
Nope. This needs to be reworked and apparently needs at least one more test